File: Python/OneLang/One/Serialization/JsonSerializer.py
from OneLangStdLib import *
import OneLang.One.Ast.AstTypes as astTypes
import OneLang.One.Ast.Interfaces as ints
import OneLang.One.Ast.Types as types
import OneLang.index as index
import re

class JsonSerializer:
    def __init__(self, lit_types):
        self.circle_detector = Map()
        self.lit_types = lit_types
    
    def pad(self, str):
        return "\n".join(list(map(lambda x: f'''    {x}''', re.split("\\n", str))))
    
    def serialize(self, obj):
        decl_type = obj.get_declared_type()
        if obj.is_null():
            return "null"
        elif astTypes.TypeHelper.equals(decl_type, self.lit_types.string):
            return JSON.stringify(obj.get_string_value())
        elif astTypes.TypeHelper.equals(decl_type, self.lit_types.boolean):
            return "true" if obj.get_boolean_value() else "false"
        elif astTypes.TypeHelper.is_assignable_to(decl_type, self.lit_types.array):
            items = []
            for item in obj.get_array_items():
                items.append(self.serialize(item))
            return "[]" if len(items) == 0 else f'''[\n{self.pad(",\n".join(items))}\n]'''
        elif astTypes.TypeHelper.is_assignable_to(decl_type, self.lit_types.map):
            items = []
            for key in obj.get_map_keys():
                value = obj.get_map_value(key)
                items.append(f'''"{key}": {self.serialize(value)}''')
            return "{}" if len(items) == 0 else f'''{{\n{self.pad(",\n".join(items))}\n}}'''
        elif isinstance(decl_type, astTypes.ClassType) or isinstance(decl_type, astTypes.InterfaceType):
            raw_value = obj.get_unique_identifier()
            if self.circle_detector.has(raw_value):
                return f'''{{"$ref":"{self.circle_detector.get(raw_value)}"}}'''
            id = f'''id_{self.circle_detector.size}'''
            self.circle_detector.set(raw_value, id)
            
            value_type = obj.get_value_type()
            decl = (decl_type).get_decl()
            
            members = []
            
            members.append(f'''"$id": "{id}"''')
            
            if value_type != None and not astTypes.TypeHelper.equals(value_type, decl_type):
                members.append(f'''"$type": "{value_type.repr()}"''')
            
            for field in list(filter(lambda x: not x.is_static, decl.fields)):
                if "json-ignore" in field.attributes:
                    continue
                #console.log(`processing ${field.parentInterface.name}::${field.name}`);
                value = obj.get_field(field.name)
                serialized_value = self.serialize(value)
                if not serialized_value in ["[]", "{}", "null", "false", "\"\""]:
                    members.append(f'''"{field.name}": {serialized_value}''')
            return "{}" if len(members) == 0 else f'''{{\n{self.pad(",\n".join(members))}\n}}'''
        elif isinstance(decl_type, astTypes.EnumType):
            return f'''"{obj.get_enum_value_as_string()}"'''
        return "\"<UNKNOWN-TYPE>\""

File: Java/src/main/java/OneLang/One/Serialization/JsonSerializer/JsonSerializer.java
package OneLang.One.Serialization.JsonSerializer;

import OneLang.One.Ast.AstTypes.ClassType;
import OneLang.One.Ast.AstTypes.EnumType;
import OneLang.One.Ast.AstTypes.IInterfaceType;
import OneLang.One.Ast.AstTypes.InterfaceType;
import OneLang.One.Ast.AstTypes.TypeHelper;
import OneLang.One.Ast.Interfaces.IType;
import OneLang.One.Ast.Types.Class;
import OneLang.One.Ast.Types.LiteralTypes;
import OneLang.One.Ast.Types.Package;
import OneStd.ReflectedValue;

import java.util.Map;
import OneLang.One.Ast.Types.LiteralTypes;
import java.util.LinkedHashMap;
import OneStd.RegExp;
import java.util.Arrays;
import java.util.stream.Collectors;
import OneLang.One.Ast.Interfaces.IType;
import OneStd.JSON;
import java.util.ArrayList;
import OneLang.One.Ast.AstTypes.ClassType;
import OneLang.One.Ast.AstTypes.InterfaceType;
import OneLang.One.Ast.AstTypes.IInterfaceType;
import OneLang.One.Ast.Types.Field;
import java.util.List;
import OneLang.One.Ast.AstTypes.EnumType;
import OneStd.ReflectedValue;

public class JsonSerializer {
    public Map<Object, String> circleDetector;
    public LiteralTypes litTypes;
    
    public JsonSerializer(LiteralTypes litTypes)
    {
        this.litTypes = litTypes;
        this.circleDetector = new LinkedHashMap<Object, String>();
    }
    
    public String pad(String str) {
        return Arrays.stream(Arrays.stream(str.split("\\n", -1)).map(x -> "    " + x).toArray(String[]::new)).collect(Collectors.joining("\n"));
    }
    
    public String serialize(ReflectedValue obj) {
        var declType = ((IType)obj.getDeclaredType());
        if (obj.isNull())
            return "null";
        else if (TypeHelper.equals(declType, this.litTypes.string))
            return JSON.stringify(obj.getStringValue());
        else if (TypeHelper.equals(declType, this.litTypes.boolean_))
            return obj.getBooleanValue() ? "true" : "false";
        else if (TypeHelper.isAssignableTo(declType, this.litTypes.array)) {
            var items = new ArrayList<String>();
            for (var item : obj.getArrayItems())
                items.add(this.serialize(item));
            return items.size() == 0 ? "[]" : "[\n" + this.pad(items.stream().collect(Collectors.joining(",\n"))) + "\n]";
        }
        else if (TypeHelper.isAssignableTo(declType, this.litTypes.map)) {
            var items = new ArrayList<String>();
            for (var key : obj.getMapKeys()) {
                var value = obj.getMapValue(key);
                items.add("\"" + key + "\": " + this.serialize(value));
            }
            return items.size() == 0 ? "{}" : "{\n" + this.pad(items.stream().collect(Collectors.joining(",\n"))) + "\n}";
        }
        else if (declType instanceof ClassType || declType instanceof InterfaceType) {
            var rawValue = obj.getUniqueIdentifier();
            if (this.circleDetector.containsKey(rawValue))
                return "{\"$ref\":\"" + this.circleDetector.get(rawValue) + "\"}";
            var id = "id_" + this.circleDetector.size();
            this.circleDetector.put(rawValue, id);
            
            var valueType = ((IType)obj.getValueType());
            var decl = (((IInterfaceType)declType)).getDecl();
            
            var members = new ArrayList<String>();
            
            members.add("\"$id\": \"" + id + "\"");
            
            if (valueType != null && !TypeHelper.equals(valueType, declType))
                members.add("\"$type\": \"" + valueType.repr() + "\"");
            
            for (var field : Arrays.stream(decl.getFields()).filter(x -> !x.getIsStatic()).toArray(Field[]::new)) {
                if (field.getAttributes().containsKey("json-ignore"))
                    continue;
                //console.log(`processing ${field.parentInterface.name}::${field.name}`);
                var value = obj.getField(field.getName());
                var serializedValue = this.serialize(value);
                if (!new ArrayList<>(List.of("[]", "{}", "null", "false", "\"\"")).stream().anyMatch(serializedValue::equals))
                    members.add("\"" + field.getName() + "\": " + serializedValue);
            }
            return members.size() == 0 ? "{}" : "{\n" + this.pad(members.stream().collect(Collectors.joining(",\n"))) + "\n}";
        }
        else if (declType instanceof EnumType)
            return "\"" + obj.getEnumValueAsString() + "\"";
        return "\"<UNKNOWN-TYPE>\"";
    }
}